backend/src/main/kotlin/app/ehrenamtskarte/backend/verification/webservice/schema/CardMutationService.kt

```diff
@@ -26,48 +26,54 @@ class CardMutationService {
             val targetedRegionId = card.regionId
             if (!Authorizer.mayCreateCardInRegion(user, targetedRegionId)) {
                 throw UnauthorizedException()
             }
             val activationSecret =
-                card.activationSecretBase64?.let {
-                    Base64.getDecoder().decode(card.activationSecretBase64)
-                }
+                    card.activationSecretBase64?.let {
+                        val decodedRawActivationSecret = Base64.getDecoder().decode(it)
+                        CardActivator.hashActivationSecret(decodedRawActivationSecret)
+                    }
 
             CardRepository.insert(
-                Base64.getDecoder().decode(card.cardInfoHashBase64),
-                activationSecret,
-                card.cardExpirationDay,
-                card.regionId,
-                user.id.value,
-                card.codeType
+                    Base64.getDecoder().decode(card.cardInfoHashBase64),
+                    activationSecret,
+                    card.cardExpirationDay,
+                    card.regionId,
+                    user.id.value,
+                    card.codeType
             )
         }
         return true
     }
 
     @GraphQLDescription("Activate a dynamic entitlement card")
     fun activateCard(
-        project: String,
-        cardInfoHashBase64: String,
-        activationSecretBase64: String,
-        overwrite: Boolean,
-        dfe: DataFetchingEnvironment
+            project: String,
+            cardInfoHashBase64: String,
+            activationSecretBase64: String,
+            overwrite: Boolean,
+            dfe: DataFetchingEnvironment
     ): CardActivationResultModel {
         val context = dfe.getContext<GraphQLContext>()
         val projectConfig =
-            context.backendConfiguration.projects.find { it.id == project }
-                ?: throw NullPointerException("Project not found")
+                context.backendConfiguration.projects.find { it.id == project }
+                        ?: throw NullPointerException("Project not found")
         val cardHash = Base64.getDecoder().decode(cardInfoHashBase64)
-        val activationSecret = Base64.getDecoder().decode(activationSecretBase64)
-        val card = transaction {
-            CardRepository.findByHashAndActivationSecret(project, cardHash, activationSecret)
+        val rawActivationSecret = Base64.getDecoder().decode(activationSecretBase64)
+        val card = transaction { CardRepository.findByHash(project, cardHash) }
+        val activationSecretHash = card?.activationSecretHash
+
+        System.out.println(activationSecretHash?.size)
+        System.out.println(rawActivationSecret.size)
+        if (card == null || activationSecretHash == null) {
+            return CardActivationResultModel(ActivationState.failed)
         }
 
-        if (card == null) {
+        if (CardActivator.verifyActivationSecret(rawActivationSecret, activationSecretHash)) {
```

Comment:
Should be inverted I think?

Reply:
Ahh. Makes sense. Then just invert and it should work.